File: crates/web/ui/src/tool-lifecycle.ts
import type {
	AssistantHistoryMessage,
	ContextBudgetMetadata,
	ToolLifecycleEvent,
	ToolLifecyclePayload,
	ToolResult,
} from "./types/ws-events";

export interface ToolInvocationSnapshot {
	lifecycle: ToolLifecycleEvent;
	runId?: string;
	executionMode?: string;
	messageIndex?: number;
	assistantMessage?: AssistantHistoryMessage;
	assistantMessageIndex?: number;
	contextBudget?: ContextBudgetMetadata;
	accumulatedArguments?: string;
}

export interface ToolInvocationMetadata {
	runId?: string;
	executionMode?: string;
	messageIndex?: number;
	assistantMessage?: AssistantHistoryMessage;
	assistantMessageIndex?: number;
	contextBudget?: ContextBudgetMetadata;
	accumulatedArguments?: string;
}

export interface TerminalToolPresentation {
	success: boolean;
	rejected: boolean;
	result: ToolResult | string | null;
	error: string | null;
}

const TOOL_LIFECYCLE_STAGES = new Set([
	"created",
	"input_streaming",
	"input_ready",
	"waiting_for_execution",
	"executing",
	"execution_progress",
	"result_ready",
	"completed",
	"rejected",
	"cancelled",
]);

function isRecord(value: unknown): value is Record<string, unknown> {
	return value !== null && typeof value === "object" && !Array.isArray(value);
}

function isArguments(value: unknown): value is Record<string, unknown> {
	return isRecord(value);
}

function isLifecycleResult(value: unknown): value is string | null {
	return value === null || typeof value === "string";
}

function hasLifecycleBase(value: Record<string, unknown>): boolean {
	return (
		typeof value.toolCallId === "string" &&
		value.toolCallId.length > 0 &&
		typeof value.toolName === "string" &&
		Number.isSafeInteger(value.sequence) &&
		Number.isSafeInteger(value.emittedAtMs) &&
		(value.runId === undefined || typeof value.runId === "string") &&
		(value.contextBudget === undefined || isRecord(value.contextBudget)) &&
		typeof value.stage === "string" &&
		TOOL_LIFECYCLE_STAGES.has(value.stage)
	);
}

export function isToolLifecycleEvent(value: unknown): value is ToolLifecycleEvent {
	if (!(isRecord(value) && hasLifecycleBase(value))) return false;
	switch (value.stage) {
		case "created":
			return value.providerIndex === null || Number.isSafeInteger(value.providerIndex);
		case "input_streaming":
			return typeof value.argumentsDelta === "string";
		case "input_ready":
		case "waiting_for_execution":
			return isArguments(value.arguments);
		case "executing":
			return isArguments(value.arguments) && Number.isSafeInteger(value.startedAtMs);
		case "execution_progress":
			return isArguments(value.arguments) && Number.isSafeInteger(value.elapsedMs) && typeof value.message === "string";
		case "result_ready":
		case "completed":
			return (
				isArguments(value.arguments) &&
				typeof value.success === "boolean" &&
				isLifecycleResult(value.result) &&
				(value.error === null || typeof value.error === "string")
			);
		case "rejected":
			return isArguments(value.arguments) && typeof value.reason === "string" && typeof value.result === "string";
		case "cancelled":
			return (value.arguments === null || isArguments(value.arguments)) && typeof value.reason === "string";
		default:
			return false;
	}
}

export function isToolLifecyclePayload(value: unknown): value is ToolLifecyclePayload {
	return isRecord(value) && value.state === "tool_lifecycle" && isToolLifecycleEvent(value);
}

export function toToolLifecycleEvent(lifecycle: ToolLifecycleEvent): ToolLifecycleEvent {
	const base = {
		toolCallId: lifecycle.toolCallId,
		toolName: lifecycle.toolName,
		sequence: lifecycle.sequence,
		emittedAtMs: lifecycle.emittedAtMs,
		runId: lifecycle.runId,
		contextBudget: lifecycle.contextBudget,
	};
	switch (lifecycle.stage) {
		case "created":
			return { ...base, stage: lifecycle.stage, providerIndex: lifecycle.providerIndex };
		case "input_streaming":
			return {
				...base,
				stage: lifecycle.stage,
				argumentsDelta: lifecycle.argumentsDelta,
			};
		case "input_ready":
		case "waiting_for_execution":
			return { ...base, stage: lifecycle.stage, arguments: lifecycle.arguments };
		case "executing":
			return {
				...base,
				stage: lifecycle.stage,
				arguments: lifecycle.arguments,
				startedAtMs: lifecycle.startedAtMs,
			};
		case "execution_progress":
			return {
				...base,
				stage: lifecycle.stage,
				arguments: lifecycle.arguments,
				elapsedMs: lifecycle.elapsedMs,
				message: lifecycle.message,
			};
		case "result_ready":
		case "completed":
			return {
				...base,
				stage: lifecycle.stage,
				arguments: lifecycle.arguments,
				success: lifecycle.success,
				result: lifecycle.result,
				error: lifecycle.error,
			};
		case "rejected":
			return {
				...base,
				stage: lifecycle.stage,
				arguments: lifecycle.arguments,
				reason: lifecycle.reason,
				result: lifecycle.result,
			};
		case "cancelled":
			return {
				...base,
				stage: lifecycle.stage,
				arguments: lifecycle.arguments,
				reason: lifecycle.reason,
			};
	}
}

function definedMetadata(metadata: ToolInvocationMetadata): ToolInvocationMetadata {
	return Object.fromEntries(
		Object.entries(metadata).filter(([, value]) => value !== undefined),
	) as ToolInvocationMetadata;
}

export function reduceToolInvocation(
	previous: ToolInvocationSnapshot | undefined,
	lifecycle: ToolLifecycleEvent,
	metadata: ToolInvocationMetadata = {},
): ToolInvocationSnapshot {
	const normalized = toToolLifecycleEvent(lifecycle);
	const nextMetadata = definedMetadata(metadata);
	if (
		previous &&
		previous.lifecycle.toolCallId === normalized.toolCallId &&
		previous.lifecycle.sequence > normalized.sequence
	) {
		return { ...previous, ...nextMetadata };
	}
	const accumulatedArguments =
		nextMetadata.accumulatedArguments ??
		(normalized.stage === "input_streaming"
			? `${previous?.accumulatedArguments ?? ""}${normalized.argumentsDelta}`
			: previous?.accumulatedArguments);
	return {
		...previous,
		...nextMetadata,
		accumulatedArguments,
		lifecycle: normalized,
	};
}

export function toolInvocationKey(sessionKey: string, runId: string | undefined, toolCallId: string): string {
	return `${sessionKey}:${runId || ""}:${toolCallId}`;
}

export function toolLifecycleArguments(lifecycle: ToolLifecycleEvent, accumulatedArguments?: string): unknown {
	switch (lifecycle.stage) {
		case "created":
			return undefined;
		case "input_streaming": {
			const argumentsText = accumulatedArguments ?? lifecycle.argumentsDelta;
			try {
				return JSON.parse(argumentsText) as unknown;
			} catch {
				return argumentsText;
			}
		}
		case "input_ready":
		case "waiting_for_execution":
		case "executing":
		case "execution_progress":
		case "result_ready":
		case "completed":
		case "rejected":
			return lifecycle.arguments;
		case "cancelled":
			return lifecycle.arguments ?? undefined;
	}
}

export function isTerminalToolLifecycle(lifecycle: ToolLifecycleEvent): boolean {
	return lifecycle.stage === "completed" || lifecycle.stage === "rejected" || lifecycle.stage === "cancelled";
}

function decodeToolResult(result: string | null): ToolResult | string | null {
	if (result === null) return null;
	try {
		const parsed: unknown = JSON.parse(result);
		return isRecord(parsed) ? (parsed as ToolResult) : result;
	} catch {
		return result;
	}
}

export function terminalToolPresentation(lifecycle: ToolLifecycleEvent): TerminalToolPresentation | null {
	switch (lifecycle.stage) {
		case "completed":
			return {
				success: lifecycle.success,
				rejected: false,
				result: decodeToolResult(lifecycle.result),
				error: lifecycle.error,
			};
		case "rejected":
			return {
				success: false,
				rejected: true,
				result: decodeToolResult(lifecycle.result),
				error: lifecycle.reason,
			};
		case "cancelled":
			return {
				success: false,
				rejected: false,
				result: null,
				error: lifecycle.reason,
			};
		default:
			return null;
	}
}

File: crates/common/src/tool_lifecycle.rs
use {
    crate::ContextBudgetMetadata,
    serde::{Deserialize, Serialize},
    serde_json::Value,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToolLifecycleStage {
    Created,
    InputStreaming,
    InputReady,
    WaitingForExecution,
    Executing,
    ExecutionProgress,
    ResultReady,
    Completed,
    Rejected,
    Cancelled,
}

impl ToolLifecycleStage {
    #[must_use]
    pub const fn is_terminal(self) -> bool {
        matches!(self, Self::Completed | Self::Rejected | Self::Cancelled)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolLifecycleEvent {
    pub tool_call_id: String,
    pub tool_name: String,
    pub sequence: u64,
    pub emitted_at_ms: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub run_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_budget: Option<ContextBudgetMetadata>,
    #[serde(flatten)]
    pub update: ToolLifecycleUpdate,
}

impl ToolLifecycleEvent {
    #[must_use]
    pub const fn stage(&self) -> ToolLifecycleStage {
        self.update.stage()
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(
    tag = "stage",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ToolLifecycleUpdate {
    Created {
        provider_index: Option<usize>,
    },
    InputStreaming {
        arguments_delta: String,
    },
    InputReady {
        arguments: Value,
    },
    WaitingForExecution {
        arguments: Value,
    },
    Executing {
        arguments: Value,
        started_at_ms: u64,
    },
    ExecutionProgress {
        arguments: Value,
        elapsed_ms: u64,
        message: String,
    },
    ResultReady {
        arguments: Value,
        success: bool,
        result: Option<String>,
        error: Option<String>,
    },
    Completed {
        arguments: Value,
        success: bool,
        result: Option<String>,
        error: Option<String>,
    },
    Rejected {
        arguments: Value,
        reason: String,
        result: String,
    },
    Cancelled {
        arguments: Option<Value>,
        reason: String,
    },
}

impl ToolLifecycleUpdate {
    #[must_use]
    pub const fn stage(&self) -> ToolLifecycleStage {
        match self {
            Self::Created { .. } => ToolLifecycleStage::Created,
            Self::InputStreaming { .. } => ToolLifecycleStage::InputStreaming,
            Self::InputReady { .. } => ToolLifecycleStage::InputReady,
            Self::WaitingForExecution { .. } => ToolLifecycleStage::WaitingForExecution,
            Self::Executing { .. } => ToolLifecycleStage::Executing,
            Self::ExecutionProgress { .. } => ToolLifecycleStage::ExecutionProgress,
            Self::ResultReady { .. } => ToolLifecycleStage::ResultReady,
            Self::Completed { .. } => ToolLifecycleStage::Completed,
            Self::Rejected { .. } => ToolLifecycleStage::Rejected,
            Self::Cancelled { .. } => ToolLifecycleStage::Cancelled,
        }
    }
}

/// Latest authoritative state of a tool invocation in an active agent run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActiveToolInvocation {
    #[serde(flatten)]
    pub lifecycle: ToolLifecycleEvent,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accumulated_arguments: Option<String>,
    /// Current budget retained internally so cancellation can publish terminal metadata.
    #[serde(skip)]
    pub context_budget: Option<ContextBudgetMetadata>,
}

impl ActiveToolInvocation {
    #[must_use]
    pub fn arguments(&self) -> Option<&Value> {
        match &self.lifecycle.update {
            ToolLifecycleUpdate::InputReady { arguments }
            | ToolLifecycleUpdate::WaitingForExecution { arguments }
            | ToolLifecycleUpdate::Executing { arguments, .. }
            | ToolLifecycleUpdate::ExecutionProgress { arguments, .. }
            | ToolLifecycleUpdate::ResultReady { arguments, .. }
            | ToolLifecycleUpdate::Completed { arguments, .. }
            | ToolLifecycleUpdate::Rejected { arguments, .. } => Some(arguments),
            ToolLifecycleUpdate::Created { .. }
            | ToolLifecycleUpdate::InputStreaming { .. }
            | ToolLifecycleUpdate::Cancelled {
                arguments: None, ..
            } => None,
            ToolLifecycleUpdate::Cancelled {
                arguments: Some(arguments),
                ..
            } => Some(arguments),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn context_budget() -> ContextBudgetMetadata {
        ContextBudgetMetadata {
            context_window: 200_000,
            max_input_tokens: 180_000,
            max_output_tokens: 20_000,
            compaction_ratio: 90,
            prompt_tokens: 1_000,
            tool_schema_tokens: 200,
            available_input_tokens: 179_800,
            compaction_budget: 161_820,
            usage_percent: 1,
            compaction_required: false,
        }
    }

    #[test]
    fn lifecycle_event_round_trips_with_authoritative_metadata() -> Result<(), serde_json::Error> {
        let event = ToolLifecycleEvent {
            tool_call_id: "call-1".to_owned(),
            tool_name: "overwrite_file".to_owned(),
            sequence: 4,
            emitted_at_ms: 1_234,
            run_id: Some("run-1".to_owned()),
            context_budget: Some(context_budget()),
            update: ToolLifecycleUpdate::ExecutionProgress {
                arguments: serde_json::json!({"path": "/tmp/report.md"}),
                elapsed_ms: 2_000,
                message: "wait for result [2] sec.".to_owned(),
            },
        };

        let value = serde_json::to_value(&event)?;
        assert_eq!(value["stage"], "execution_progress");
        assert_eq!(value["toolCallId"], "call-1");
        assert_eq!(value["runId"], "run-1");
        assert_eq!(value["contextBudget"]["contextWindow"], 200_000);
        assert_eq!(value["arguments"]["path"], "/tmp/report.md");
        assert_eq!(value["elapsedMs"], 2_000);

        let decoded: ToolLifecycleEvent = serde_json::from_value(value)?;
        assert_eq!(decoded, event);
        assert_eq!(decoded.stage(), ToolLifecycleStage::ExecutionProgress);
        assert!(!decoded.stage().is_terminal());
        Ok(())
    }

    #[test]
    fn active_snapshot_round_trips_streamed_input_and_progress() -> Result<(), serde_json::Error> {
        let streamed = ActiveToolInvocation {
            lifecycle: ToolLifecycleEvent {
                tool_call_id: "call-stream".to_owned(),
                tool_name: "overwrite_file".to_owned(),
                sequence: 2,
                emitted_at_ms: 2_000,
                run_id: Some("run-1".to_owned()),
                context_budget: None,
                update: ToolLifecycleUpdate::InputStreaming {
                    arguments_delta: "report".to_owned(),
                },
            },
            execution_mode: None,
            accumulated_arguments: Some(r#"{"content":"report"#.to_owned()),
            context_budget: None,
        };
        let progress = ActiveToolInvocation {
            lifecycle: ToolLifecycleEvent {
                tool_call_id: "call-progress".to_owned(),
                tool_name: "execute_command".to_owned(),
                sequence: 5,
                emitted_at_ms: 5_000,
                run_id: Some("run-1".to_owned()),
                context_budget: None,
                update: ToolLifecycleUpdate::ExecutionProgress {
                    arguments: serde_json::json!({"command": "sleep 10"}),
                    elapsed_ms: 10_000,
                    message: "wait for result [10] sec.".to_owned(),
                },
            },
            execution_mode: Some("sandbox".to_owned()),
            accumulated_arguments: None,
            context_budget: None,
        };

        let streamed_value = serde_json::to_value(&streamed)?;
        assert_eq!(
            streamed_value["accumulatedArguments"],
            r#"{"content":"report"#
        );
        assert_eq!(streamed_value["runId"], "run-1");
        let progress_value = serde_json::to_value(&progress)?;
        assert_eq!(progress_value["elapsedMs"], 10_000);
        assert_eq!(progress_value["executionMode"], "sandbox");
        assert_eq!(
            progress.arguments(),
            Some(&serde_json::json!({"command": "sleep 10"}))
        );
        Ok(())
    }
}
